src/main/java/jenkins/plugins/git/maintenance/TaskExecutor.java

```diff
@@ -0,0 +1,113 @@
+import java.util.logging.Logger;
+
+public class TaskExecutor implements Runnable {
+
+    Task maintenanceTask;
+    List<GitMaintenanceSCM.Cache> caches;
+
+    private static final Logger LOGGER = Logger.getLogger(TaskExecutor.class.getName());
+
+    public TaskExecutor(Task maintenanceTask){
+        this.maintenanceTask = new Task(maintenanceTask);
+        caches = getCaches();
+    }
+
+    @Override
+    public void run() {
+
+        LOGGER.log(Level.FINE,"Running maintenance task " + maintenanceTask.getTaskName() + " on git caches.");
+        GitClient gitClient;
+        for(GitMaintenanceSCM.Cache cache : caches){
+
+            // For now adding lock to all kinds of maintenance tasks. Need to study on which task needs a lock and which doesn't.
+            Lock lock = cache.getLock();
+            File cacheFile = cache.getCacheFile();
+            try {
+                gitClient = getGitClient(cacheFile);
+                if(gitClient == null)
+                    return;
+
+                TaskType taskType = maintenanceTask.getTaskType();
+
+                lock.lock();
+                LOGGER.log(Level.FINE,"Cache " + cacheFile.getName() + " locked.");
+
+                executeMaintenanceTask(gitClient,taskType);
+
+            } catch (IOException | InterruptedException e) {
+                // What is this error???
+                LOGGER.log(Level.WARNING,"Git Client couldn't be initialized.");
+            }finally {
+                lock.unlock();
+                LOGGER.log(Level.FINE,"Cache " + cacheFile.getName() + " unlocked.");
+            }
+        }
+
+    }
+
+    void executeMaintenanceTask(GitClient gitClient,TaskType taskType) throws InterruptedException{
+
+        if(taskType.equals(TaskType.GC)){
+            gitClient.maintenance("gc");
+        }else if(taskType.equals(TaskType.COMMIT_GRAPH)){
+            gitClient.maintenance("commit-graph");
+        }else if(taskType.equals(TaskType.PREFETCH)){
+            gitClient.maintenance("prefetch");
+        }else if(taskType.equals(TaskType.INCREMENTAL_REPACK)){
+            gitClient.maintenance("incremental-repack");
+        }else if(taskType.equals(TaskType.LOOSE_OBJECTS)){
+            gitClient.maintenance("loose-objects");
+        }else{
+            LOGGER.log(Level.WARNING,"Invalid maintenance task.");
+        }
+
+    }
+
+    private List<Integer> getGitVersion(){
+        return MaintenanceTaskConfiguration.getGitVersion();
+    }
+
+    List<GitMaintenanceSCM.Cache> getCaches(){
+        return GitMaintenanceSCM.getCaches();
+    }
+
+    GitClient getGitClient(File file) throws IOException, InterruptedException {
+        // What exactly is default tool here?
+        TaskListener listener = new LogTaskListener(LOGGER,Level.FINE);
+        final Jenkins jenkins = Jenkins.getInstanceOrNull();
+        // How to get Jenkins controller as the node?
+        GitTool gitTool = GitUtils.resolveGitTool(null,jenkins,null, listener);
```

Comment:
do you want the default git installation? 
`GitTool.getDefaultInstallation()` gives the default installation